docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+   existing standard tools and libraries, including command-line use for
+   testing and experimentation.
+
+### Definitions
+
+- The **event stream** of a node is a single, time-ordered, heterogeneous
+  stream of event items.
+
+- Each **event item** comprises an **event datum** (for example, block header
+  metadata for a new-block event), and zero or more optional **events**.
+
+- An **event** means the [ABCI `Event` data type][abci-event], which comprises
+  a string type and zero or more string key-value **event attributes**.
+
+  The use of the new terms "event item" and "event datum" is to avert confusion
+  between the values that are published to the event bus (what we call here
+  "event items") and the ABCI `Event` data type.
+
+- The node assigns each event item a unique identifier string called a
+  **cursor**.  A cursor must be unique among all events published by a single
+  node, but it is not required to be unique globally across nodes.
+
+  Cursors are time-ordered so that given event items A and B, if A was
+  published before B, then cursor(A) < cursor(B) in lexicographic order.
+
+  A minimum viable cursor implementation is a tuple consisting of a timestamp
+  and a sequence number (e.g., `16CCC798FB5F4670-0123`). However, it may also
+  be useful to append basic type information to a cursor, to allow efficient
+  filtering (e.g., `16CCC87E91869050-0091:BeginBlock`).
+
+  The initial implementation will use the minimum viable format.
+
+### Discussion
+
+The node maintains an **event log**, a shared ordered record of the events
+published to its event bus within an operator-configurable time window.  The
+initial implementation will store the event log in-memory, and the operator
+will be given two configuration settings.  Note, these names are provisional:
+
+- `[event-subscription] time-window`: A duration before present during which the
+  node will retain event items published. Setting this value to zero disables
+  event subscription.
+
+- `[event-subscription] max-items`: A maximum number of event items that the
+  node will retain within the time window. If the number of items exceeds this
+  value, the node discardes the oldest items in the window. Setting this value
+  to zero means that no limit is imposed on the number of items.
+
+The node will retain all events within the time window, provided they do not
+exceed the maximum number.  These config parameters allow the operator to
+loosely regulate how much memory and storage the node allocates to the event
+log.
+
+The event log is shared among all subscribers to the node.
+
+> **Open question:** Should events persist across node restarts?
+>
+> The current event API does not persist events. Some events may be _indexed_,
+> but the node does not keep a complete record: Only BeginBlock and EndBlock
+> events are passed to the indexer, if an indexer is configured at all.
+>
+> The initial implementation will not persist events across restarts.
```

Comment:
It would make things easier if they were not. Nodes seldom need to restart: only for upgrades and server maintenance. But if things are not persisted, we're creating a pain point around reliability for clients which will need to be programmed to find another node to receive the missing events from. In the case of coordinated upgrades, if a client falls behind it will never be able to catch up because all nodes will stop at around the same time.

Reply:
Yes, it's a lot easier if we don't persist events across restarts.

As for clients "catching up", apart from restarts that seems to me like an operator concern. If the application needs a durable event stream, the operator should configure a long enough time window that a client should get everything with high-enough probability.

I consider true "durability" of events to be outside the scope of this design. If a client so depends on a complete picture of the event stream that it would have to probe multiple nodes to get it, then I'd argue there's a separate design problem there about long-term storage of events.

Reply:
Is there any need to consider the retry semantics of indexers? Do we persist those events?

Reply:
> Is there any need to consider the retry semantics of indexers? Do we persist those events?

I'm not sure what you mean about retry semantics. I don't think the indexer has any retry logic, at least not from what I've seen in KV and PSQL. Have I missed something?

Reply:
I think there is a CLI for reindexing events: https://github.com/tendermint/tendermint/blob/master/cmd/tendermint/commands/reindex_event.go

Reply:
I just realised that this also uses the persisted ABCI responses so I was wrong - the persisted events do actually get used somewhere

Reply:
Oh, I see, I thought you meant an automatic retry mechanism.

Yes, I think this tool probably implies we can't (yet) get rid of the event data that we're persisting in the block/state store. The indexer relies on the ABCI events in particular. But for RPC purposes, I don't think there's any way to do that replay.

Reply:
This is a slight digression but if you aren't running a tx indexer you probably shouldn't be storing events (they take up an annoying amount of storage).

I guess the point I was trying to articulate is that if we are persisting events for the indexer then why can't we reuse them for the events system however I acknowledge that they are relatively different sets (or one is a subset of the other) so this probably doesn't make sense.

Reply:
> This is a slight digression but if you aren't running a tx indexer you probably shouldn't be storing events (they take up an annoying amount of storage).

That's a very good point. Though I wonder—if someone decides to restart their node to add an indexer, and wants to use that replay tool, would they then be out of luck? 

Maybe that's fine. In any case, I would be happy to look into that problem separately.

> [I]f we are persisting events for the indexer then why can't we reuse them for the events system

Yeah, I very much wanted to do that, but we both save too much (cf. your point above) and too little (only indexable events).

We may at some point want to offload event data to disk, e.g., if the memory pressure becomes a problem at the time scales operators want. If we do that, though, we probably want it to be a separate table from the blockstore/statestore, since its pruning requirements are different.

That said—if we _do_ make that change, we could probably use that opportunity to rip out the event data from the other stores.

Reply:
> I think that's a reasonable behaviour to expect from a node. If you switch on the indexer you shouldn't be able to replay events before you turned it on

That makes sense to me. And while this is a tangent as you said, I wonder: If we have that expectation, do we really need to be storing all those old event data in the block/state store? If we don't expect nodes to support replay, could we drop the tool _and_ the data?

Reply:
> That said—if we do make that change, we could probably use that opportunity to rip out the event data from the other stores.

> do we really need to be storing all those old event data in the block/state store? If we don't expect nodes to support replay, could we drop the tool and the data?

The `TxSearch` API currently rely on the indexer. If we drop the data, it might broken the API?

Reply:
> The `TxSearch` API currently rely on the indexer. If we drop the data, it might make the API broken?

You raise a good question. The answer, I think, is complicated.

The search API only works if the KV indexer is enabled. But it doesn't use the event data from the state store, it has its own log of events.  Conversely, even if the events are in the state store, if the KV indexer isn't enabled search will not be possible.

I'll have to do more research, but it looks right now like the only thing using the event data from the state store is the reindex tool.

Reply:
I added another paragraph to the **Future Work** section about this, and updated this paragraph to confirm that we're not going to worry about persisting the event log in the initial implementation.